File: Downloads/Dashboard_Project_Testing-Updated (2)/Dashboard_Project_Testing-Updated/frontend/src/app/core/services/data-versioning.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import { tap, catchError } from 'rxjs/operators';
import { of } from 'rxjs';
import { environment } from '@env/environment';

export interface DataVersion {
  versionId: string;
  schemaId: string;
  schemaName: string;
  tableName: string;
  checksum: string;
  rowCount: number;
  uploadedAt: string;
  fileName: string;
  versionNumber: number;
  isDuplicate: boolean;
  originalVersionId?: string;
}

export interface VersionCheckResult {
  isDuplicate: boolean;
  existingVersion?: DataVersion;
  newChecksum: string;
}

@Injectable({
  providedIn: 'root'
})
export class DataVersioningService {
  private apiUrl = `${environment.apiUrl}/data/versions`;
  private versions$ = new BehaviorSubject<DataVersion[]>([]);

  constructor(private http: HttpClient) {}

  /**
   * Calculate checksum for uploaded data
   */
  calculateChecksum(data: any[]): string {
    const dataStr = JSON.stringify(data);
    return this.simpleHash(dataStr);
  }

  /**
   * Check if data is a duplicate of an existing version
   */
  checkDuplicate(schemaId: string, checksum: string): Observable<VersionCheckResult> {
    return this.http.post<VersionCheckResult>(
      `${this.apiUrl}/check-duplicate`,
      { schemaId, checksum }
    ).pipe(
      catchError((error) => {
        console.error('Error checking duplicate:', error);
        return of({ isDuplicate: false, newChecksum: checksum });
      })
    );
  }

  /**
   * Register a new data version
   */
  registerVersion(
    schemaId: string,
    schemaName: string,
    tableName: string,
    checksum: string,
    rowCount: number,
    fileName: string,
    isDuplicate: boolean = false,
    originalVersionId?: string
  ): Observable<DataVersion> {
    return this.http.post<DataVersion>(
      `${this.apiUrl}/register`,
      {
        schemaId,
        schemaName,
        tableName,
        checksum,
        rowCount,
        fileName,
        isDuplicate,
        originalVersionId
      }
    ).pipe(
      tap((version) => {
        const current = this.versions$.value;
        this.versions$.next([...current, version]);
      }),
      catchError((error) => {
        console.error('Error registering version:', error);
        throw error;
      })
    );
  }

  /**
   * Get all versions for a schema
   */
  getVersionsBySchema(schemaId: string): Observable<DataVersion[]> {
    return this.http.get<DataVersion[]>(`${this.apiUrl}/schema/${schemaId}`).pipe(
      tap((versions) => this.versions$.next(versions)),
      catchError((error) => {
        console.error('Error fetching versions:', error);
        return of([]);
      })
    );
  }

  /**
   * Get specific version details
   */
  getVersion(versionId: string): Observable<DataVersion> {
    return this.http.get<DataVersion>(`${this.apiUrl}/${versionId}`).pipe(
      catchError((error) => {
        console.error('Error fetching version:', error);
        throw error;
      })
    );
  }

  /**
   * Get all versions observable
   */
  getVersions$(): Observable<DataVersion[]> {
    return this.versions$.asObservable();
  }

  /**
   * Simple hash function for checksum (can be replaced with crypto if needed)
   */
  private simpleHash(str: string): string {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    return Math.abs(hash).toString(16);
  }
}

File: backend/src/main/java/com/example/dashboard_backend/service/DataVersioningService.java
package com.example.dashboard_backend.service;

import com.example.dashboard_backend.ingestion.metadata.DataVersioningRepository;
import com.example.dashboard_backend.model.DataVersion;
import com.example.dashboard_backend.model.VersionCheckResult;
import com.example.dashboard_backend.model.CheckDuplicateRequest;
import com.example.dashboard_backend.model.RegisterVersionRequest;
import org.springframework.stereotype.Service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Optional;

/**
 * Manages data versioning with checksum-based deduplication.
 * Prevents duplicate uploads from creating redundant versions,
 * tracks version metadata, and enables version comparison in dashboards.
 */
@Service
public class DataVersioningService {

    private static final Logger logger = LoggerFactory.getLogger(DataVersioningService.class);
    private final DataVersioningRepository repository;

    public DataVersioningService(DataVersioningRepository repository) {
        this.repository = repository;
    }

    /**
     * Check if data with this checksum already exists for the schema.
     * Returns existing version if duplicate found, otherwise indicates new data.
     */
    public VersionCheckResult checkDuplicate(CheckDuplicateRequest request) {
        logger.info("Checking for duplicate: schema={}, checksum={}", request.schemaId(), request.checksum());

        Optional<DataVersion> existing = repository.findByChecksumAndSchema(request.schemaId(), request.checksum());

        if (existing.isPresent()) {
            logger.info("Duplicate found: version={}, schemaId={}", existing.get().versionId(), request.schemaId());
            return new VersionCheckResult(true, existing.get(), request.checksum());
        }

        logger.info("No duplicate found, new data: checksum={}", request.checksum());
        return new VersionCheckResult(false, null, request.checksum());
    }

    /**
     * Register a new data version with checksum and metadata.
     * If isDuplicate=true, originalVersionId points to the reused version.
     */
    public DataVersion registerVersion(RegisterVersionRequest request) {
        logger.info("Registering version: schema={}, isDuplicate={}, filename={}",
                request.schemaId(), request.isDuplicate(), request.fileName());

        DataVersion version = repository.registerVersion(
                request.schemaId(),
                request.schemaName(),
                request.tableName(),
                request.checksum(),
                request.rowCount(),
                request.fileName(),
                request.isDuplicate(),
                request.originalVersionId(),
                request.createdBy()
        );

        logger.info("Version registered: id={}, number={}, rows={}",
                version.versionId(), version.versionNumber(), version.rowCount());

        return version;
    }

    /**
     * Get all versions for a schema, ordered by version number (oldest to newest).
     */
    public List<DataVersion> getVersionsBySchema(String schemaId) {
        logger.info("Fetching versions for schema: {}", schemaId);
        List<DataVersion> versions = repository.getVersionsBySchema(schemaId);
        logger.info("Found {} versions for schema: {}", versions.size(), schemaId);
        return versions;
    }

    /**
     * Get a specific version by ID.
     */
    public Optional<DataVersion> getVersion(String versionId) {
        logger.info("Fetching version: {}", versionId);
        return repository.getVersionById(versionId);
    }

    /**
     * Delete a version (useful for cleanup or removing corrupted uploads).
     */
    public boolean deleteVersion(String versionId) {
        logger.info("Deleting version: {}", versionId);
        return repository.deleteVersion(versionId) > 0;
    }

    /**
     * Calculate SHA-256 checksum of data.
     * Note: Frontend already calculates simple hash; backend uses this for verification.
     */
    public static String calculateChecksum(String jsonData) {
        try {
            java.security.MessageDigest md = java.security.MessageDigest.getInstance("SHA-256");
            byte[] messageDigest = md.digest(jsonData.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : messageDigest) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (java.security.NoSuchAlgorithmException e) {
            logger.error("SHA-256 algorithm not available", e);
            // Fallback to simple hash
            return String.valueOf(jsonData.hashCode());
        }
    }
}
